### PR TITLE
Reduce calls to SplFileInfo::realpath() in the Magento\Setup\Module\Di\Code\Reader\ClassesScanner class

### PR DESCRIPTION
**Pull request by kschroeder**
Reduces calls to SplFileInfo::realpath() in the Magento\Setup\Module\Di\Code\Reader\ClassesScanner class.

### Description
I have had issues with fairly slow setup:di:compile calls.  In the course of my investigation, I ran the XDebug profiler and found that there were several million calls to SplFileInfo::realpath().  This change reduced my compile time from 9-10 minutes to 3 minutes.

[update]

I have been able to find several 10's of seconds (100+ seconds on my test box) of efficiency to be gained by creating a single extraction class that looks only for the first namespace/class match.

[update]

The single extraction class has been changed so that it now scans the entire PHP source file to find unconventionally defined classes as well

### Manual testing scenarios
This test simply reduces the number of system calls on the filesystem.  There is a functional change, not a logical change.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=8965) <br/>All committers have signed the CLA.

**Comment by piotrkwiecinski** [comment]
@kschroeder nice one :) I've tested in on dev setup and for me it went down from 7 to 4 min and on CI server from 2:30 to 46 sec. Much appreciated you've saved quite a few CPU cycles :).

**Comment by kschroeder** [comment]
@piotrkwiecinski Sweet, I was hoping it wasn't just me!  :-)

**Comment by schmengler** [comment]
Nice, verifiable PR with huge benefits! Unfortunately your change in `ClassesScanner.php` triggered the cyclomatic complexity check in the static test suite. This is not really a problem with your change, but rather with the original method. Let's take this as a chance to refactor the method a bit more. A simple **extract method** in the loop should do it.

**Comment by kschroeder** [comment]
@schmengler I updated the PR with a new commit that separates the extract() method.  If this one doesn't pass I'll separate the $classNames loop.

**Comment by schmengler** [comment]
```
--------------------------------------------------------------------------------
FOUND 1 ERROR(S) AFFECTING 1 LINE(S)
--------------------------------------------------------------------------------
 59 | ERROR | Missing function doc comment
--------------------------------------------------------------------------------
```

**Comment by PieterCappelle** [comment]
The best community ever. Great PR.

**Comment by kschroeder** [comment]
Note that I found another 2 minutes or so I could squeeze out of it, which is where those new commits are coming from.  However, I found an edge case that is breaking it a little bit so it is not ready to be merged yet.

**Comment by AydinHassan** [comment]
Nice - these new improvements will probably fix #8782 also.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (schmengler) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php]: If I understand that code correctly, it will break at least the following special cases:

- `namespace X { }` syntax
- multiple classes per file

While both do not happen in the core, we have no control over third party code. Keep in mind that the Magento DI framework can also used to integrate third party libraries that are not Magento modules.

File: setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/FileClassScannerTest.php
Comment (schmengler) [inline setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/FileClassScannerTest.php]: what is the weird exception case with that file?

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (schmengler) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: To increase the chance of getting this released as soon as possible, please make new constructor arguments optional so that there is no backwards incompatibility. Then if it is null, use `ObjectManager::getInstance()->get()` to retrieve the object (yes, that's the recommended  practice)

Comment (schmengler) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: Nevermind, you already removed it :)

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (schmengler) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: the **** copyright header is always the current year...

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (schmengler) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: Is it possible to restrict that check to the relative path below the Magento root. Just in case, somebody installs Magento below a directory called "generation"? It's unlikely but you never know...

**Review by schmengler — CHANGES_REQUESTED** [review]

**Comment by schmengler** [comment]
Thank you for the update! This is neat, but I see some problems with potential edge cases. Please also check the review comments marked as "outdated", I just commented them in previous commits but they are still valid.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: heh, I need to be better about my copy and paste, will change

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: I'm working on refactoring this part of the code, there is some interaction with some of the tests somehow that did not like this.  I had intended to use DirectoryList class to get the full path but introducing that dependency broke lots of other tests so I opted to just do a string match.  I will make the check more intentional.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: I wasn't aware of that practice, I have been trying like the plague to avoid using the ObjectManager like that.

File: setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/FileClassScannerTest.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/FileClassScannerTest.php]: It is similar to #8782.  The class keyword has the same token number as the class definition keyword.  That means that the code used to determine the class name has to understand the different contexts.  I had tried to write the test so that I could mock the code but I couldn't get it to replicate.  So I just referenced the one file that was breaking.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php]: Yeah, this code, which is slightly a hack to deal with the ::class issue, expects that the code people integrate follows the most basic of Magento coding standards.  I was also hoping that I wouldn't have to iterate over the entire file.  But if this is an important enough edge case (and I will presume it is unless told otherwise) then I will scan the whole file.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (piotrkwiecinski) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: Instead of 2 nested ifs I would be better to  have them as !$isGenerated && isset

File: setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/FileClassScannerTest.php
Comment (piotrkwiecinski) [inline setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/FileClassScannerTest.php:163]: @kschroeder what do you think about extracting classes to data provider(s) here? It would be more flexible to add more cases in future. I think you could make these tests cleaner this way.

Comment (kschroeder) [reply setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/FileClassScannerTest.php:163]: Perhaps, but I was following the output of the original code so-as to minimize any downstream pain.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: Yes.  Pushing soon.

**Comment by kschroeder** [comment]
@schmengler @piotrkwiecinski I keep triggering the cyclomatic complexity warning for PHPMD.  Because of the recursive and iterative nature of parsing tokens (because of the inconsistent representation in the `token_get_all()` array), I'm hesitant to make it all perfect because those method calls do add up.  I haven't measured it recently but at one point there was a method called 55 million times.  With 55 million calls micro-optimizations do add up.  So I would like a little guidance on what you would prefer: reduce the individual method complexity or notate MD to ignore CC?

**Comment by schmengler** [comment]
> So I would like a little guidance on what you would prefer: reduce the individual method complexity or notate MD to ignore CC?

If reducing the method complexity has performance impact, I'd say go for optimized performance in this case and mark the class as ignored.

**Comment by davidalger** [comment]
> If reducing the method complexity has performance impact, I'd say go for optimized performance in this case and mark the class as ignored.

I would agree, but include a comment indicating why it's ignored, so it doesn't eventually get refactored by some random dude trying to get rid of all the cyclomatic complexity warnings. ;)

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (schmengler) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: Don't get me wrong, continue avoiding it. And please still add any dependencies to the constructor (i.e. DirectoryList), just make them optional as `DirectoryList $directoryList = null` and fall back to the object manager. This way it's fully backwards compatible - and the fallback can be removed later.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (orlangur) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php:71]: This condition is really confusing.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (orlangur) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: Why do we need `require_once` here and not `require` since we follow PSR-2?

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (orlangur) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php]: Better use PHP Parser.

File: setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/ClassesScannerTest.php
Comment (orlangur) [inline setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/ClassesScannerTest.php:27]: Why need `realpath` here?

File: setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/FileClassScannerTest.php
Comment (orlangur) [inline setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/FileClassScannerTest.php]: PHPDoc shall start with `/**`, please fix all occurrences.

**Review by orlangur — CHANGES_REQUESTED** [review]

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (orlangur) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: All new properties shall be `private`.

**Review by orlangur — CHANGES_REQUESTED** [review]

**Comment by orlangur** [comment]
> a single extraction class that looks only for the first namespace/class match

Does it mean the FQCN matching class itself? Why we cannot obtain it from file path itself?

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php:71]: It simply excludes the generation directory from the local result cache if it exists.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: Presumably because of the possibility of an incorrect class definition in one file affecting another operation.  This is is how it was done in the original code.  https://github.com/magento/magento2/blob/364635691c8baa5cc30d58c1c1879fc309f5b048/setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php]: Why?

File: setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/ClassesScannerTest.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/ClassesScannerTest.php:27]: To get a clear indication that it is the path that is broken if there is an error during debugging.

File: setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/FileClassScannerTest.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/FileClassScannerTest.php]: Will do

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: Will do

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (orlangur) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: I noticed `require_once` derived from old code, of course.

> possibility of an incorrect class definition

But why not fail fast in such case? I'm not saying such change must be a part of your PR, just asking if you see some logic in `_once` as you are already in context of this code.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (orlangur) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php]: Because there is no need to reinvent the wheel. Proposed implementation is quite low-level and hard to maintain while this problem is easily solvable with PHP Parser (I did some prototyping with it and FQCN resolution in particular is shown in documentation).

But please clarify https://github.com/magento/magento2/pull/8965#issuecomment-289475460 first - I thought this `extract` is intended to find all class references at first glance.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (orlangur) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php:71]: I mean it should have been `if (!$realPath)` ;)

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (orlangur) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: Please remove all such empty lines. I'm not sure it is enforced by static tests currently but is definitely inconsistent with Magento 2 codebase.

**Review by orlangur — CHANGES_REQUESTED** [review]

**Review by schmengler — APPROVED** [review]

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (maghamed) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: it's better to compare with null than to make instanceof opeartion.
It's better from teh performance point of view

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (maghamed) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php]: newly created method should be private , not protected

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (maghamed) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php]: please remove this blank lines

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (maghamed) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php:67]: we forbid SuppressWarning for Cyclomatic complexity in newly created code

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (maghamed) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php]: don't use protected for newly created methods

**Review by maghamed — CHANGES_REQUESTED** [review]

**Comment by orlangur** [comment]
@kschroeder @schmengler 
> With 55 million calls micro-optimizations do add up. So I would like a little guidance on what you would prefer: reduce the individual method complexity or notate MD to ignore CC?

> If reducing the method complexity has performance impact ...

Performance impact needs to be demonstrated first.

I believe implementation based on PHP Parser would be both better maintainable and has better performance. But before doing this please clarify https://github.com/magento/magento2/pull/8965#issuecomment-289475460

**Comment by schmengler** [comment]
@orlangur Regarding #8965, I explicitly requested to not do this. Quoting my review:

> If I understand that code correctly, it will break at least the following special cases:
> 
> - `namespace X { }` syntax
> - multiple classes per file
> 
> While both do not happen in the core, we have no control over third party code. Keep in mind that the Magento DI framework can also used to integrate third party libraries that are not Magento modules.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (schmengler) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php]: @orlangur Since micro optimization is a thing here, I would not introduce a full fledged parser until it's proven that the impact is negligible. Both should have linear complexity, but the current solution is just one loop that does not exactly what's needed and not more.

**Comment by orlangur** [comment]
@schmengler I understand this, but do we need to support PSR-0,1,4 incompatible third-party code?

Some 99.99% of libraries would not need source code parsing, for other proper wrappers could be created so that we do not scan bad smelling code ever.

**Comment by schmengler** [comment]
@orlangur I prefer a "no surprises" approach here. I've been in the 0.01 percent group before in a similar case. Debugging issues in the compiler and finding workarounds is a pain. Could you write wrappers for these libraries? Sure. If you *know* that this is the problem. But it will be hidden very well and we have no reliable way to warn the user about it.

**Comment by orlangur** [comment]
@schmengler makes sense to me, thanks, I'm not even sure all libraries currently used directly in core are PSR-0,1,4 compliant.

Added TODO for myself to check later for possibility of this kind of improvement keeping developer experience seamless.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (orlangur) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php]: > micro optimization is a thing here

This needs to be shown in numbers... For the cyclomatic complexity, not for alternative solution with third-party parser.

>  I would not introduce a full fledged parser until it's proven that the impact is negligible

Ok, no objections :) I was just asking to evaluate possibility of PHP Parser usage here and now, it can be improved later iteratively.

**Review by orlangur — COMMENTED** [review]
"general feedback without explicit approval" :)

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php:71]: Oh, well, you'd have to ask the original author of the class.  I didn't write that line.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (kschroeder) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php:67]: See https://github.com/magento/magento2/pull/8965#issuecomment-289130082

**Comment by schmengler** [comment]
@maghamed the PR still needs your approval, please leave feedback. This is an important change and it would be a shame if we don't get it into the next release.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php
Comment (maghamed) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php:39]: it's not allowed to make function calls in the constructor, just assignments to class variables.

here we receive  $directoryList and after that make a call to get Path

File: setup/src/Magento/Setup/Module/Di/Code/Reader/InvalidFileException.php
Comment (maghamed) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/InvalidFileException.php:5]: Custom Exceptions are not recommended to be introduced in each module.

Here is more discussion on this subject -  https://github.com/magento/magento2/pull/9315/files#r112514206

File: setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/ClassesScannerTest.php
Comment (maghamed) [inline setup/src/Magento/Setup/Test/Unit/Module/Di/Code/Reader/ClassesScannerTest.php:22]: empty line

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (maghamed) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php:67]: > > So I would like a little guidance on what you would prefer: reduce the individual method complexity or notate MD to ignore CC?

> If reducing the method complexity has performance impact, I'd say go for optimized performance in this case and mark the class as ignored.

@schmengler  Performance and Complexity are two different metrics. And in this specific case these metrics do not correlate. Because high Cyclomatic complexity doesn't lead to high performance and vice versa.

In Magento we have a rule, that any new code should not have SuppressWarnings for CouplingBetween objects and CyclomaticComplexity. Because it's from the beginning has a Technical Debt. 

High CyclomaticComplexity often says that some functionality has not been fully covered with automated testing. Because it's too hard to cover all the possible conditional logic and exit points for this method. 

Moreover such code is fragile and really hard to maintain in future.

Some time ago we even actively used [CRAP](https://testing.googleblog.com/2011/02/this-code-is-crap.html) metrics for this.

So, this method should be decomposed on some smaller once

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (maghamed) [inline setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php:47]: magento constructors should contain just assigments, and no business logic allowed inside

**Review by maghamed — CHANGES_REQUESTED** [review]

**Comment by kschroeder** [comment]
Forget this.  I'm done.  I have about a thousand other things I need to be working on.

**Comment by mzeis** [comment]
@kschroeder Thank you for putting so much work into this pull request! I hope we still can get your changes into core because - as @schmengler said - it would be a shame to not have it in future Magento versions.

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (mzeis) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php:47]: @maghamed I'd like to argue that this is a valid use case according to 2.3.1 of the [Technical Guidelines](http://devdocs.magento.com/guides/v2.1/coding-standards/technical-guidelines/technical-guidelines.html):

> 2.3.1. Constructor SHOULD throw an exception when validation of an argument has failed.

In the example for 2.3.1, an array of options is validated and an `InvalidArgumentException` is thrown if one of the objects in the array doesn't implement a particular interface.

In the given code the provided filename has to be an existing file so for me, this is a valid validation. Which alternative approach would you suggest?

File: setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php
Comment (mzeis) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php:67]: @maghamed See https://github.com/magento/magento2/pull/8965#issuecomment-289122432 for the explanation by @kschroeder why there might be a performance impact when decomposing this into more methods. How can we proceed here? Should we try to get hard numbers on the difference in performance?

File: setup/src/Magento/Setup/Module/Di/Code/Reader/InvalidFileException.php
Comment (mzeis) [reply setup/src/Magento/Setup/Module/Di/Code/Reader/InvalidFileException.php:5]: For `Magento\Setup\Module\Di\Code\Reader\FileClassScanner::__construct` I'd suggest to go with `\Magento\Framework\Exception\FileSystemException`. It's used the same way in `\Magento\Framework\Filesystem\Directory\Write::assertIsFile`.

For `Magento\Setup\Module\Di\Code\Reader\FileClassScanner::isBracedNamespace`, I'm a bit undecided which of the existing exeptions is the best fit. Maybe `\Magento\Framework\Exception\NotFoundException`, at least for one of the two cases?

**Comment by joshuaadickerson** [comment]
@kschroeder I was about to jump to your defense on rewriting that method and I still think it should be accepted and then it can be reworked at any point in the future. Then I checked out the class and I looked at your argument. Taking a method and rewriting it to use internal (to the class) methods, even at 55 million calls, should not make any noticeable difference. I think all of those variables you're declaring in the extract() method are properties of the class. When you call extract(), they represent the state of that object. The class's only purpose is for that method (which is good).

I don't care to submit a PR or follow the coding guidelines, but I created a test of what it would look like and how that would affect performance: https://gist.github.com/joshuaadickerson/240f3633d99106fb7aed44a98a0ef0fd

It would be interesting to see what the difference between your class and my class are for the build that you said requires 55m calls.

**Comment by mzeis** [comment]
@maghamed Please can you have a look at the comments I gave on April 23rd? Thanks!

**Comment by antonkril** [comment]
The PR improves the state of the product. Let's merge.

**Review by antonkril — APPROVED** [review]

**Pull request merged**
